rclcpp/src/rclcpp/parameter_client.cpp

```diff
@@ -273,6 +273,58 @@ AsyncParametersClient::set_parameters_atomically(
   return future_result;
 }
 
+std::shared_future<std::vector<rcl_interfaces::msg::SetParametersResult>>
+AsyncParametersClient::delete_parameters(
+  const std::vector<std::string> & parameters_names)
+{
+  std::vector<rclcpp::Parameter> parameters;
+  for (const std::string & name : parameters_names) {
+    parameters.push_back(rclcpp::Parameter(name));
+  }
+  auto future_result = set_parameters(parameters);
+
+  return future_result;
+}
+
+std::shared_future<std::vector<rcl_interfaces::msg::SetParametersResult>>
+AsyncParametersClient::load_parameters(
+  const std::string & yaml_filename)
+{
+  // parse yaml file
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  rcl_params_t * rcl_parameters = rcl_yaml_node_struct_init(allocator);
+  const char * path = yaml_filename.c_str();
+  if (!rcl_parse_yaml_file(path, rcl_parameters)) {
+    throw std::runtime_error("Unable to get get parameters from yaml file");
+  }
+
+  // create list of parameters to set
+  rclcpp::ParameterMap parameter_map = rclcpp::parameter_map_from(rcl_parameters);
+  std::vector<rclcpp::Parameter> parameters;
+  std::string fully_qualified_name =
+    node_topics_interface_->get_node_base_interface()->get_fully_qualified_name();
+
+  // only add remote_node parameters
+  for (auto & params : parameter_map) {
+    std::string node_name = params.first;
+    if (node_name == fully_qualified_name ||
+      node_name == "/**" ||
+      (node_name.substr(node_name.find("/*/") + 3) == remote_node_name_))
```

Comment:
Can you explain the logic of this?

The first case is like "if it matches the name of _this_ node", the second case is "if it is a wild card" (makes sense), and the third is "if the name of the _remote_ node matches it, allowing for `/*/` at the beginning"?

What's confusing to me is that it seems like one condition applies to this node, and the last one applies to a remote node (also the logic there is kind of confusing if you don't have expert knowledge of the wildcard syntax).

So I guess I'm checking my understanding and asking for some comments for each condition that describe what they're doing (even if it seems obvious to you).

Reply:
I mixed the two nodes...
Want we want to do is to check witch parameters are for the target node `remote_node_name_`.
To do so we have to check the yaml `node_name` there is 3 options:

1. namespace/node_name 
2. `/**` wild card
3. `/*/` node_name (any namespace)

I made a mistake on the first it should be `node_name == remote_node_name_`.
The second is ok.
The third must compare what is after `/*/` and the remote_node_name_ without namespace!

Hence to fit 1. we need `remote_node_name_` to be fully qualified and to fit 3. we need to extract the node name from it.

Another issue is when the ParameterClient is a client to its own node, remote_node_name_ isn't the fully_qualified_name.

https://github.com/ros2/rclcpp/blob/0088e35052270eca4ccbb0f0c2325f4030c39baf/rclcpp/src/rclcpp/parameter_client.cpp#L42-L46

To fix it, one way is to break the API by adding the remote_node_namespace param. The other is to use `remote_node_name_ = node_base_interface->get_fully_qualified_name()` that may change the behavior of some implementations.

What is your opinion?

Reply:
> I made a mistake on the first it should be node_name == remote_node_name_.

👍 I thought so, but I just wanted to double check.

> The other is to use remote_node_name_ = node_base_interface->get_fully_qualified_name() that may change the behavior of some implementations.

I'd go with this. I don't think it should be an issue, because the behavior before seems wrong to me, so anyone trying to use it with namespaces may have not gotten it working at all.

I'd prefer this over adding an explicit node namespace argument, because I believe the intention was always that the node name given to the client was meant to be the fully qualified name (something we should have documented and tested, but apparently haven't).

That's my impression, but I didn't think deeply about it. You may have to try it and see.

Reply:
Done